// bold-ui/lib/components/PivotTable/PivotTableCell/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pivotTableCellCreateStyles = void 0;
var utils_1 = require("./utils");
var model_1 = require("./model");
exports.pivotTableCellCreateStyles = function (theme, key, maxValue, _a) {
    var types = _a.types, children = _a.children;
    var isValue = types.has(model_1.PivotTableCellType.VALUE);
    var isValueOrEmpty = isValue || types.has(model_1.PivotTableCellType.EMPTY);
    var isHeaderOrTotal = types.has(model_1.PivotTableCellType.HEADER) || types.has(model_1.PivotTableCellType.TOTAL);
    var isOnlyValue = isValue && types.size === 1;
    var _b = utils_1.calculateCellColor(theme, isOnlyValue, maxValue, children), color = _b.color, backgroundColor = _b.backgroundColor;
    return {
        root: {
            gridArea: key,
            backgroundColor: backgroundColor,
            color: color,
            borderTop: "1px solid " + theme.pallete.divider,
            borderLeft: "1px solid " + theme.pallete.divider,
            display: 'flex',
            justifyContent: isValueOrEmpty ? 'flex-end' : 'flex-start',
            fontWeight: (isHeaderOrTotal ? 'bold' : 'normal'),
            alignItems: 'center',
            width: '100%',
            height: '100%',
            padding: '0.5rem 1rem',
        },
        endRowBorder: {
            borderBottom: "1px solid " + theme.pallete.divider,
        },
        endColumnBorder: {
            borderRight: "1px solid " + theme.pallete.divider,
        },
    };
};
